File: tree.cpp
#include <iostream>
using namespace std;

// class Node{
//     public:
//         int value;
//         Node* left;
//         Node* right;
//         Node(int value){
//             this->value = value;
//             this->right = nullptr;
//             this->left = nullptr;
//         };
// };

class Tree{
    
    struct Node
    {
        int value;
        Node* right;
        Node* left;
    };
    Node* root;

    public:
        Node* getRoot(){
            return this->root;
        };
        void addNode(int value){
            Node* newNode = new Node();
            newNode->left = nullptr;
            newNode->right = nullptr;
            newNode->value = value;
            if (root == nullptr)
                root = newNode;
            else{
                Node* nodePtr = getRoot();
                while (nodePtr != nullptr)
                {
                    if (value < nodePtr->value){
                        if (nodePtr->left != nullptr)
                            nodePtr = nodePtr->left;
                        else{
                            nodePtr->left = newNode;      
                            return; 
                        }
                    }
                        
                    if (newNode->value > nodePtr->value){
                        if (nodePtr->right != nullptr)
                            nodePtr = nodePtr->right;
                        else{
                            nodePtr->right = newNode;        
                            return;
                        }
                    }
                }  
            }
        };

        void printInOrder(Node* nodePtr){
            if (nodePtr == nullptr)
                return;
            else{
                cout << nodePtr->value << endl;
                printInOrder(nodePtr->left);
                printInOrder(nodePtr->right);
            }
            
        }
};



int main(){
    Tree myTree = Tree();
    myTree.addNode(10);
    myTree.addNode(9);
    myTree.addNode(11);
    myTree.addNode(15);
    myTree.addNode(46);
    myTree.addNode(52);
    myTree.addNode(91);
    myTree.addNode(100);



    // myTree.printInOrder(myTree.getRoot());
    int a = 2^2;
    cout << a;
}

File: tree.js
class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class Tree {
    constructor() {
        this.root = null;
    }

    getRoot() {
        return this.root;
    }

    insert(newNode) {
        if (this.root === null) {
            this.root = newNode;
        }
        else {
            let nodePtr = this.root;
            while (nodePtr != null) {
                if (newNode.value <= nodePtr.value) {
                    if (nodePtr.left != null) {
                        nodePtr = nodePtr.left;
                    }
                    else {
                        nodePtr.left = newNode;
                        return;
                    }
                }
                if (newNode.value >= nodePtr.value) {
                    if (nodePtr.right != null) {
                        nodePtr = nodePtr.right;
                    }
                    else {
                        nodePtr.right = newNode;
                        return;
                    }
                }
            }
        }
    }

    recursiveInsert(newNode, nodePtr){
        if(this.root === null)
            this.root = newNode;
        else{
            if (newNode.value <= nodePtr.value) {
                if (nodePtr.left != null) {
                    this.recursiveInsert(newNode, nodePtr.left);
                }
                else {
                    nodePtr.left = newNode;
                    return;
                }
            }
            if (newNode.value >= nodePtr.value) {
                if (nodePtr.right != null) {
                    this.recursiveInsert(newNode, nodePtr.right);
                }
                else {
                    nodePtr.right = newNode;
                    return;
                }
            }
        }
    }

    printInOrder(nodePtr) {
        if (nodePtr === null)
            return;
        else{
            this.printInOrder(nodePtr.left);
            console.log(nodePtr.value);
            this.printInOrder(nodePtr.right);
            return;
        }
    }
    printPreOrder(nodePtr) {
        if (nodePtr === null)
            return;
        else{
            console.log(nodePtr.value);
            this.printPreOrder(nodePtr.left);
            this.printPreOrder(nodePtr.right);
            return;
        }
    }
    printPostOrder(nodePtr) {
        if (nodePtr === null)
            return;
        else{
            this.printPostOrder(nodePtr.left);
            this.printPostOrder(nodePtr.right);
            console.log(nodePtr.value);
            return;
        }
    }
}

let myTree = new Tree();
// myTree.insert(new Node(10));
// myTree.insert(new Node(8));
// myTree.insert(new Node(11));
// myTree.insert(new Node(22));
// myTree.insert(new Node(12));

myTree.recursiveInsert(new Node(10), myTree.getRoot());
myTree.recursiveInsert(new Node(8), myTree.getRoot());
myTree.recursiveInsert(new Node(11), myTree.getRoot());
myTree.recursiveInsert(new Node(22), myTree.getRoot());
myTree.recursiveInsert(new Node(12), myTree.getRoot());
myTree.recursiveInsert(new Node(7), myTree.getRoot());











myTree.printPostOrder(myTree.getRoot());